pkg/agent/iptables/iptables.go

```diff
@@ -154,6 +322,89 @@ func (c *Client) ensureRule(table string, chain string, ruleSpec []string) error
 	return nil
 }
 
+// storeRule saves rule for reconcile
+func (c *Client) storeRule(table string, chain string, ruleSpec []string) {
+	var chains chainRules = nil
+	if val, ok := c.store.Load(table); !ok {
+		chains = make(map[string][][]string)
+	} else {
+		chains = val.(chainRules)
+	}
+	if _, ok := chains[chain]; !ok {
+		chains[chain] = nil
+	}
+	chains[chain] = append(chains[chain], ruleSpec)
+	c.store.Store(table, chains)
+}
+
+// PrintStoredRules dumps stored ip rules for debugging.
+func (c *Client) PrintStoredRules() string {
+	tables := []string{
+		NATTable,
+		FilterTable,
+		MangleTable,
+		RawTable,
+	}
+	wr := bytes.NewBuffer(nil)
+	for _, table := range tables {
+		wr.Write([]byte(table + ":\n"))
+		val, ok := c.store.Load(table)
+		if !ok {
+			wr.Write([]byte("not found\n"))
+			continue
+		}
+		chains := val.(chainRules)
+		wr.Write([]byte(fmt.Sprintf("size: %d\n", len(chains))))
+		for chain, ruleSpecs := range chains {
+			wr.Write([]byte(chain + ":\n"))
+			for _, ruleSpec := range ruleSpecs {
+				wr.Write([]byte(fmt.Sprintf("len=%d ", len(ruleSpec))))
+				wr.Write([]byte(strings.Join(ruleSpec, " ") + "\n"))
+			}
+			wr.Write([]byte("\n"))
+		}
+	}
+	return wr.String()
+}
+
+// compareRuleSpec returns true if two rule spec are the same.
+func compareRuleSpec(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	// by content
+	visited := make(map[string]interface{}, len(a))
+	for _, s := range a {
+		visited[strings.ToLower(s)] = nil
+	}
+	for _, s := range b {
+		if _, ok := visited[strings.ToLower(s)]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// IsInRuleStore returns true if rule is in rule store.
+func (c *Client) IsInRuleStore(table string, chain string, ruleSpec []string) bool {
+	var chains chainRules = nil
+	val, ok := c.store.Load(table)
+	if !ok {
+		return false
+	}
+	chains = val.(chainRules)
+	if _, ok := chains[chain]; !ok {
+		return false
+	}
+
+	for _, storeRule := range chains[chain] {
+		if compareRuleSpec(storeRule, ruleSpec) {
+			return true
+		}
+	}
+	return false
+}
+
```

Comment:
Perhaps using iptables-restore operations like kube-proxy will make the processing simpler, but not needed to be in this PR of course.

Reply:
I have not looked into Kube-proxy, not sure how was iptable-save, iptable-restore was used in their context. Of course there is always opportunity to improve.